backend/src/org/commcare/resources/model/ResourceTable.java

```diff
@@ -298,100 +404,69 @@ public void prepareResources(ResourceTable master, CommCareInstance instance)
     /**
      * Makes some (or all) of the table's resources available
      *
-     * @param master       The global resource to prepare against. Used to establish whether resources need to be fetched
-     *                     remotely
+     * @param master       The global resource to prepare against. Used to
+     *                     establish whether resources need to be fetched remotely
      * @param instance     The instance to prepare against
-     * @param toInitialize The ID of a single resource after which the table preparation can stop.
-     * @throws UnresolvedResourceException       If a resource could not be identified and is required
-     * @throws UnfullfilledRequirementsException If some resources are incompatible with the current version of CommCare
+     * @param toInitialize The ID of a single resource after which the table
+     *                     preparation can stop.
+     * @throws UnresolvedResourceException       Required resource couldn't be
+     *                                           identified
+     * @throws UnfullfilledRequirementsException resource(s) incompatible with
+     *                                           current CommCare version
      */
-    public void prepareResources(ResourceTable master, CommCareInstance instance, String toInitialize)
+    public void prepareResources(ResourceTable master,
+                                 CommCareInstance instance,
+                                 String toInitialize)
             throws UnresolvedResourceException, UnfullfilledRequirementsException {
-        Stack<Resource> v = GetUnreadyResources();
-        int round = -1;
-        while (!v.isEmpty() &&
-                (toInitialize == null ||
-                        this.getResourceWithId(toInitialize).getStatus() == Resource.RESOURCE_STATUS_UNINITIALIZED)) {
-            round++;
+
+        boolean idNeedsInitialization = true;
+        if (toInitialize != null) {
+            Resource res = this.getResourceWithId(toInitialize);
+            if (res != null && res.getStatus() != Resource.RESOURCE_STATUS_UNINITIALIZED) {
+                idNeedsInitialization = false;
+            }
+        }
```

Comment:
Might not be valid to do this upfront, out of the loop. If the status of toInitialize gets updated in the loop.